packages/fuels-abigen-macro/tests/test_projects/tuples/out/debug/tuples-abi.json

```diff
@@ -0,0 +1,20 @@
+[
+  {
+    "type": "function",
+    "inputs": [
+      {
+        "name": "input",
+        "type": "(u64, u64)",
```

Comment:
Maybe test with 4 different types eg `(u64, bool)` and `(u32, u8)` just to be extra safe?

Reply:
I love correct software as much as the next person, but if you look closely at the tuple handling code:

```Rust
pub fn parse_tuple_param(param: &Property) -> Result<ParamType, Error> {
  let mut params: Vec<ParamType> = Vec::new();

  let mut chars = param.type_field.chars();
  chars.next(); // Remove "("
  chars.next_back(); // Remove ")"

  for ele in chars.as_str().split(',') {
      let tuple_type = Property {
          name: "".to_owned(),
          type_field: ele.trim().to_owned(),
          components: None,
      };
      params.push(parse_param(&tuple_type)?);
  }

  Ok(ParamType::Tuple(params))
}
```

After we strip `'('` and `')'`, it goes through regular/primitive `ParamType` parsing: `units`, `bool`, etc., that's been tested extensively. 

The same applies to your comment [here](https://github.com/FuelLabs/fuels-rs/pull/205#discussion_r843551352). Where now, instead of encoding, we're talking about decoding:

```Rust
ParamType::Tuple(types) => {
    let mut tokens = vec![];
    let mut new_offset = offset;

    for t in types {
        let res = self.decode_param(t, data, new_offset)?;
        new_offset = res.new_offset;
        tokens.push(res.token);
    }

    let result = DecodeResult {
        token: Token::Tuple(tokens),
        new_offset,
    };

    Ok(result)
}
```

Which, as you said in that comment, the implementation is fairly straightforward, just going through each inner type and doing ` let res = self.decode_param(t, data, new_offset)?;`, which has been fairly tested already.

As much as I _love_ tests, I personally refrain from writing unnecessary and/or redundant code (code == liability).

Reply:
Makes sense, I often forget that code (even tests) is a liability!